File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/regionList/RegionListActivity.java
package practicaltask.riseapps.com.practicaltask.ui.regionList;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import practicaltask.riseapps.com.practicaltask.PracticalTaskApp;
import practicaltask.riseapps.com.practicaltask.R;
import practicaltask.riseapps.com.practicaltask.data.enums.Region;
import practicaltask.riseapps.com.practicaltask.ui.base.BaseActivity;
import practicaltask.riseapps.com.practicaltask.ui.countryList.CountryListActivity;
import practicaltask.riseapps.com.practicaltask.ui.regionList.adapter.RegionListAdapter;

import static practicaltask.riseapps.com.practicaltask.ui.countryList.CountryListActivity.REGION_NAME_KEY;

public class RegionListActivity extends BaseActivity implements RegionListView {

    @BindView(R.id.region_recycler_view)
    protected RecyclerView regionsRecyclerView;

    private RegionListAdapter adapter;

    @Inject
    protected RegionListPresenter mainPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        PracticalTaskApp.getInstance().getApplicationComponent().inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_region_list);
        ButterKnife.bind(this);

        //Указано на всех формах "back", будем возвращать )
        initHeaderButton();
        setTitle(getString(R.string.region_list_header));

        regionsRecyclerView.setHasFixedSize(true);
        regionsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new RegionListAdapter(mainPresenter::openRegion);
        regionsRecyclerView.setAdapter(adapter);

        mainPresenter.setView(this);
    }

    @Override
    public void showRegionList(List<Region> regions) {
        adapter.setmDataset(regions);
    }

    @Override
    public void openCountryListOfRegion(Region region) {
        Intent intent = new Intent(RegionListActivity.this, CountryListActivity.class);
        intent.putExtra(REGION_NAME_KEY, region.getCodeName());
        startActivity(intent);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mainPresenter.setView(null);
    }
}

File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/regionList/RegionListActivity.kt
package practicaltask.riseapps.com.practicaltask.ui.regionList

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import butterknife.ButterKnife
import kotlinx.android.synthetic.main.activity_region_list.*
import practicaltask.riseapps.com.practicaltask.PracticalTaskApp
import practicaltask.riseapps.com.practicaltask.R
import practicaltask.riseapps.com.practicaltask.data.enums.Region
import practicaltask.riseapps.com.practicaltask.ui.base.BaseActivity
import practicaltask.riseapps.com.practicaltask.ui.countryList.CountryListActivity
import practicaltask.riseapps.com.practicaltask.ui.countryList.CountryListActivity.Companion.REGION_NAME_KEY
import practicaltask.riseapps.com.practicaltask.ui.regionList.adapter.RegionListAdapter
import javax.inject.Inject

class RegionListActivity : BaseActivity(), RegionListView {

    lateinit var adapter: RegionListAdapter

    @Inject
    lateinit var mainPresenter: RegionListPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        PracticalTaskApp.getInstance().applicationComponent.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_region_list)
        ButterKnife.bind(this)

        //Указано на всех формах "back", будем возвращать )
        initHeaderButton()
        setTitle(getString(R.string.region_list_header))

        region_recycler_view.setHasFixedSize(true)
        region_recycler_view.layoutManager = LinearLayoutManager(this)
        adapter = RegionListAdapter { mainPresenter.openRegion(it) }
        region_recycler_view.adapter = adapter

        mainPresenter.setView(this)
    }

    override fun showRegionList(regions: List<Region>) {
        adapter.setmDataset(regions)
    }

    override fun openCountryListOfRegion(region: Region) {
        val intent = Intent(this@RegionListActivity, CountryListActivity::class.java)
        intent.putExtra(REGION_NAME_KEY, region.codeName)
        startActivity(intent)
    }

    override fun onDestroy() {
        super.onDestroy()
        mainPresenter.setView(null)
    }
}
